HW8

TatyanaV24 opened the pull request.

ios-review commented on Swift-homework-main/DZ 6/SecondViewController.swift:
```diff
@@ -0,0 +1,39 @@
+    @IBOutlet weak var resultLabel: UILabel!
+    
+      
+    @IBAction func enterNum(_ sender: Any) {
```
Параметр sender все еще не используется и присутствует в сигнатуре функции.

Прочитать раздел The Target-Action Mechanism по ссылке https://developer.apple.com/documentation/uikit/uicontrol#1943645
Упростить сигнатуры функций с учетом статьи.

@IBAction функции могут иметь три типа сигнатуры:

1. Без параметров
```
 @IBAction func doAction() {
 }
```
2. С одним параметром
```
 @IBAction func doAction(_ sender: Any) {
 }
```
3. С двумя параметрами
```
 @IBAction func doAction(_ sender: Any, forEvent event: UIEvent) {
 }
```
Рекомендую использовать тот тип сигнатуры, который наиболее подходит под задачу.
Например, если не будут использоваться параметры, то можно их не указывать.

Если изменяется сигнатура функции, то по сути создается новая функция.
Новую функцию требуется заново связать с событием в сториборде, иначе приложение упадет.

ios-review commented on Swift-homework-main/DZ 6/SecondViewController.swift:
```diff
@@ -0,0 +1,39 @@
+    
+      
+    @IBAction func enterNum(_ sender: Any) {
+        guard let num = Int(numTextField.text!) else {
```
Все еще используется восклицательный знак.
Извлекать значения можно последовательно в одном guard.

```
guard let numText = numTextField.text, let num = Int(numText) else {
    resultLabel.text! = "Введите целое число в строку"
    return
}
```

Если хотя бы одно значение не будет извлечено, то код пойдет по пути else, так как не извлечение значение равносильно условию false.
Также запятая образует логическое И.

ios-review commented on Swift-homework-main/DZ 6/SecondViewController.swift:
```diff
@@ -0,0 +1,39 @@
+      
+    @IBAction func enterNum(_ sender: Any) {
+        guard let num = Int(numTextField.text!) else {
+            return resultLabel.text! = "Введите целое число в строку"
```
При установке значения не требуется его извлекать принудительно:
```
resultLabel.text = "Введите целое число в строку"
return
```

ios-review commented on Swift-homework-main/DZ 6/SecondViewController.swift:
```diff
@@ -0,0 +1,39 @@
+            return resultLabel.text! = "Введите целое число в строку"
+        }
+        var answer = 1
+        let power = Int(numTextField.text!)
```
force-unwrap - принудительное извлечение запрещено
Только в одном месте допускается восклицательный знак:
`@IBOutlet var someView: UIView! // только здесь, больше нигде`

ios-review commented on Swift-homework-main/DZ 6/SecondViewController.swift:
```diff
@@ -0,0 +1,39 @@
+        }
+        var answer = 1
+        let power = Int(numTextField.text!)
+        for _ in 1...power! {
+            answer &*= num
+        }
+        if answer < Int.max {
+            resultLabel.text = "\(String(describing: answer))"
+        } else {
+            resultLabel.text! = "Превышено допустимое значение диапазона Int. Введите другое чило."
+        }
```
Данное решение не сработает, так как оператор переполнения работает так: если значение превышает максимально допустимое, то оно становится минимально допустимым.
То есть как бы зацикливается.

Попробуйте функцию https://developer.apple.com/documentation/swift/int/2884292-multipliedreportingoverflow
Если overflow будет true, тогда можно сказать, что значение превышено.

Дополнительно попробуйте тип https://developer.apple.com/documentation/swift/double без преобразования его в Int.

ios-review commented on Swift-homework-main/DZ 6/ThirdViewController.swift:
```diff
@@ -0,0 +1,62 @@
+            let operation = Operator(rawValue: operatorString)
+            let secondOperand = Double(secondNumber.text!)!
+            
+            enum Operator: String {
+                case sum = "+"
+                case sub = "-"
+                case multi = "*"
+                case div = "/"
+                
+                func apply(_ a: Double, _ b: Double) -> Double {
+                    switch self {
+                    case .sum:
+                        return a+b
+                    case .sub:
+                        return a-b
+                    case .multi:
+                        return a*b
+                    case .div:
+                        return a/b
+                        
+                    }
+                    
+                }
+            }
```
Вынести за пределы функции.
```
enum Operator {}
class VC {}
```

ios-review commented on Swift-homework-main/DZ 6/ThirdViewController.swift:
```diff
@@ -0,0 +1,62 @@
+    
+    @IBAction func resultButton(_ sender: Any) {
+    
+        if (Double(firstNumber.text!) != nil) && (operatorTextField.text != nil) && (Double(secondNumber.text!) != nil) == true {
+            
+            let firstOperand = Double(firstNumber.text!)!
+            let operatorString = operatorTextField.text!
+            let operation = Operator(rawValue: operatorString)
+            let secondOperand = Double(secondNumber.text!)!
```
```
        guard let aText = firstNumber.text,
              let bText = secondNumber.text,
              let operatorText = operatorTextField.text,
              let a = Double(aText),
              let b = Double(bText),
              let op = Operator(rawValue: operatorText) else {
            self.resultLabel.text = "Некорректные данные"
            return
        }
```

ios-review commented on Swift-homework-main/DZ 6/ThirdViewController.swift:
```diff
@@ -0,0 +1,62 @@
+            if secondOperand != 0 {
+                self.resultLabel.text = "\(String(describing: operation?.apply(firstOperand, secondOperand)))"
+            } else {
+                self.resultLabel.text = "Делить на ноль нельзя"
```
Попробуйте не проверять.
Что получится?
Как думаете, почему?

ios-review commented on Swift-homework-main/DZ 6/ViewController.swift:
```diff
@@ -0,0 +1,26 @@
+
+class ViewController: UIViewController {
+    
+    var nameUser:[String] = []
```
теперь нужен данный массив?

ios-review commented on Swift-homework-main/DZ 6/ViewController.swift:
```diff
@@ -0,0 +1,26 @@
+    
+    
+    @IBAction func saveName(_ sender: Any) {
+        printNameLabel.text! += nameTextField.text! + " "
```
извлечь безопасно

ios-review commented on Swift-homework-main/DZ 6/ViewController.swift:
```diff
@@ -0,0 +1,26 @@
+    @IBOutlet weak var printNameLabel: UILabel!
+    
+    
+    @IBAction func saveName(_ sender: Any) {
```
параметр sender нужен?

ios-review reviewed (changes requested):
На первые два вопроса ответил в комментариях.
На третий - все 👌
Вы верно реализовали, так как и задумывалось)

Хочется, чтобы вы не стеснялись в решениях.
Я показываю всего лишь один из возможных вариантов решений.
Их бесчисленное множество)
Пробуйте, экспериментируйте.
У вас все получится!

Некоторые участки можно доработать.

ios-review commented on Swift-homework-main/DZ 6/SecondViewController.swift:
```diff
@@ -15,25 +15,24 @@ class SecondViewController: UIViewController {
+                    resultLabel.text = "Превышено допустимое значение диапазона Int. Введите другое чило."
+                }
+            }
         }
```
Следующий алгоритм будет больше соответствовать условию задачи:
 ```
   @IBAction func enterNum() {
        guard let numText = numTextField.text, let num = Int(numText) else {
            resultLabel.text = "Введите целое число в строку"
            return
        }
        
        var result: (Int, Bool) = (1, false)
        for _ in 1...num {
            result = 2.multipliedReportingOverflow(by: result.0)
            if result.1 {
                break
            }
        }
        
        if result.1 {
            resultLabel.text = "Превышено допустимое значение диапазона Int. Введите другое чило."
        } else {
            resultLabel.text = "\(String(describing: result.0))"
        }
   }
```

Требовалось возвести число 2 в степень из текстового поля.
Текущая реализация возводит введенное число в степень этого же числа.

ios-review commented on Swift-homework-main/DZ 6/SecondViewController.swift:
```diff
@@ -15,25 +15,24 @@ class SecondViewController: UIViewController {
-            answer &*= num
+            if let pow = numTextField.text, let power = Int(pow){
+            for _ in 1...power {
+                let result = answer.multipliedReportingOverflow(by: answer)
```
В программировании бизнес-приложений операторы переполнения редкоиспользуемые.
Не могу припомнить ни одного примера, когда бы они пригодились.

В программировании требовательных к производительности систем такие операторы могут пригодится, так как могут позволить написать более эффективные программы за счет эффекта переполнения.

Также базовые операторы +-/* "роняют" приложение при переполнении.
Можно использовать операторы переполнения, чтобы быть уверенными в безопасности выполнения кода программы, так как результат выполнения операторов переполнения всегда будет. Возможно он будет немного неожиданным)
Ими можно воспользоваться при работе с очень большими или очень маленькими числами на границах диапазона числового типа.

Подробнее по ссылке https://docs.swift.org/swift-book/LanguageGuide/AdvancedOperators.html#ID37

> If you try to insert a number into an integer constant or variable that cannot hold that value, by default Swift reports an error rather than allowing an invalid value to be created. This behavior gives extra safety when you work with numbers that are too large or too small.

ios-review commented on Swift-homework-main/DZ 6/ThirdViewController.swift:
```diff
@@ -19,44 +41,20 @@ class ThirdViewController: UIViewController {
             self.resultLabel.text = "Некорректные данные"
+            return
+        }
+        if b != 0 {
```
При типе Double деление можно не проверять на ноль.
Будет число infinity https://developer.apple.com/documentation/swift/double/1541006-infinity
Бесконечность.

ios-review commented on Swift-homework-main/DZ 6/ViewController.swift:
```diff
@@ -9,18 +9,18 @@ import UIKit
-    @IBAction func saveName(_ sender: Any) {
-        printNameLabel.text! += nameTextField.text! + " "
+    @IBAction func saveName() {
+        if let userName = nameTextField.text {
+            printNameLabel.text! += userName + " "
+        }
```
```
guard let enteredName = nameTextField.text, let currentNames = printNameLabel else { // Прочитали значения введенного нового имени и текущего списка имен
    return
}
printNameLabel.text = currentNames + " " userName // Записали новое значение в printNameLabel.text на базе текущего + нововведенного.
```

ios-review reviewed (approved):
Все гуд👍

Реквест можно мержить (merge).

TatyanaV24 committed:
Inital project

TatyanaV24 committed:
Задачи-блок 8

При выпоолнении задач из блока 8, появились следующие вопросы:
1. первая задача работает, но систематически вылетае, и выдает сообщение "2020-11-18 19:49:15.858414+0300 HW8[6220:446070] [TraitCollection] Class CKBrowserSwitcherViewController overrides the -traitCollection getter, which is not supported. If you're trying to override traits, you must use the appropriate API.
Fatal error: Index out of range: file Swift/ContiguousArrayBuffer.swift, line 444
2020-11-18 19:49:21.141941+0300 HW8[6220:446070] Fatal error: Index out of range: file Swift/ContiguousArrayBuffer.swift, line 444"
2. Через код смогла реализовать нужный варинат страницы, но мне неудалось расположить лейбл под картинкой с нимаенованием, как только через массив пытаюсь добавить надпись в лейбл, сами лейблы начинают очень сильно перемещаться по экрану.
3. В целом я разобралась с элементом, единственный момент. Во всех статьях и источниках, что я изучала указывают, что экраны при переходе с использованием UISegmentedControl должны содержать единый формат. По заданию, они у нас отличаются, и то ли поэтому то ли по другой причине, при переходе у меня все элеменыт сохраняются на вью. Не смогла придумать как удалять объекты при переходе по сегментам.

ios-review commented on HW8/HW8/ViewController.swift:
```diff
@@ -0,0 +1,41 @@
+        // активация кнопки назад
+        case 0:
+            self.index = (self.index >= self.catsImage.count-1) ? 0 : self.index-1
+            self.changeImageView.image = UIImage(named:catsImage[index])
```
Если нажимать вперед, а затем несколько раз назад, то приложение упадет с ошибкой:

> Fatal error: Index out of range: file Swift/ContiguousArrayBuffer.swift, line 444
> 2020-11-18 19:49:21.141941+0300 HW8[6220:446070] Fatal error: Index out of range: file Swift/ContiguousArrayBuffer.swift, line 444"

Ошибка в вычислении индекса.
Если в массиве нет элемента с запрашиваемым индексом, то будет сообщение "Index out of range", что значит "Индекс вне диапазона".
Ошибка в вычислении индекса.
Индекс 0 всегда меньше, чем self.catsImage.count-1, а значит будет выполнен код после ":", то есть `self.index - 1`.
0 - 1 = -1
-1 не валидный индекс.

TatyanaV24 replied:
Переписала условия для вывода картинок, ошибок в индексе больше нет.

ios-review commented on HW8/HW8/ViewController.swift:
```diff
@@ -0,0 +1,41 @@
+
+
+
+
```
code-style: пустые строки.
В реальной разработке очень строго относятся к Code-style.

Все любят красивое оформление)

TatyanaV24 replied:
Учту на будущее пунктик по оформлению. А есть какой-то особый список основных правил оформления кода или требования? Может где-то в документации указаны?

ios-review replied:
Есть правила к проектированию API.
https://swift.org/documentation/api-design-guidelines/

Также есть более обширные правила, большая часть которых используется во множестве команд.
https://github.com/raywenderlich/swift-style-guide

ios-review commented on HW8/HW8/ViewController.swift:
```diff
@@ -0,0 +1,41 @@
+    @IBAction func imageCollectionButton(_ sender: UIButton) {
+        switch sender.tag {
+        // активация кнопки назад
+        case 0:
```
В данном случае такой подход не является оправданным, так как количество кода не уменьшается в любом случае, а код становится более запутанным.
У другого разработчика может возникнуть вопрос: почему выполнение код различается в зависимости от tag и что значит 0, а что значит 1.

Если же выделить две функции previous и next, то другой разработчик сразу поймет, что это обработчики нажатия кнопок, для которых должны происходить соотвествующие действия.
В случае previous уменьшение индекса и обновление картинки.
В случае next увеличение индекса и обновление картинки.

В предыдущем задании так делалось, так как код был одинаковый для любой функции. Отличалось только конкретное значение модели данных, связанное с кнопками при помощи tag.

Здесь же различие именно в логике, которое проще и нагляднее показать в виде двух функций-обработчиков.

Будет здорово, если попробуете смотреть на код со стороны и задавать вопрос: "Если бы я впервые видела данный код, я бы поняла его назначение сразу?"
Если возникает вопрос: "А что значит это значение и почему от него зависит логика?", то код в большинстве случае можно сделать более понятным.

Понятность для других - супер важный критерий при написании кода ПО.

TatyanaV24 replied:
Тут все понятно, хотела уменьшить код, поэтому и объединила все в одно условие. Постараюсь думать не только о себе, но и возможных будущих коллегах.

ios-review commented on HW8/HW8/GalleryViewController.swift:
```diff
@@ -0,0 +1,91 @@
+
+class GalleryViewController: UIViewController {
+
+    let catsImage = [UIImage(named:"1"), UIImage(named:"2"),UIImage(named:"3"),UIImage(named:"4")]
```
Создать модель данных для картинки и подписи.

```
struct Cat {
    let icon: UIImage
    let name: String
}
```

Моделируйте!
Изменяемые данные группируйте по смыслу в сущности, в модели данных с помощью классов/структур/перечислений.

Композируйте простые данные в более сложные модели.
И декомпозируйте супер сложные объекты в более мелкие.

Сперва на сложном большом экране можно выделить простые повторяющиеся элементы.
В данном случае это в сумме картинка + текст.
Эти элементы располагаются сеткой 2х2.
Затем, выделив простые элементы, объединить их в модель.

Декомпозиция для выделения простых элементов -> композиция для выделения модели данных для сложного экрана.

ios-review commented on HW8/HW8/GalleryViewController.swift:
```diff
@@ -0,0 +1,91 @@
+                imageLabel.frame = CGRect(x: coordinateX, y: 700, width: Int(imageWidth), height: 30)
+                imageLabel.backgroundColor = UIColor.gray
+                //imageLabel.text = "Cute cat's"
+            }
```
Данный алгоритм можно упростить.

Если заметить, то разница для каждого элемента из массива только в frame.
Изменяется только frame для картинки+текста.
И то только x и y, а width и height постоянны.

Алгоритм в теле цикла может быть примерно такой:
1) Вычислить x и y в зависимости от index элемента.
2) Создать UIImageView с фиксированным frame.
3) Создать UILabel с фиксированным frame.
4) Создать UIView (эта вью будет контейнером для картинки и лейбла) с посчитанным x и y в пункте 1 и фиксированным width и height.
5) Добавить картинку+лейбл в вью из пункта 4.
6) Добавить вью из пункта 4 в иерархию вью для контроллера.

frame - это прямоугольник (левая верхняя вершина + размер), описывающий вью в системе координат супервью.
Так как картинка + лейбл будут добавляться в контейнер вью, то внутри нее для любой картинки и лейбла будут одинаковые frame.
То есть внутри системы координат этого контейнера frame для картинки и лейбла будут одинаковыми.

А вот frame каждого из этих четырех контейнеров в системе координат вью для контроллера будут отличаться, так как их нужно расположить сеткой 2х2.

Скрин может помочь:
<img width="444" alt="Screenshot 2020-11-19 at 19 58 56" src="https://user-images.githubusercontent.com/62285448/99669610-e4b37700-2aa1-11eb-9e0c-c27e0e772bef.png">

ios-review replied:
В iOS используется перевернутая декартова система координат.
Начало в левом верхнем углу.
Ось X направлена вправо.
Ось Y направлена вниз.

ios-review replied:
Каждая вью образует свою систему координат.
Если в нее добавляется вью, то ее frame рассматривается именно в системе координат вью, в которую ее добавили.

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        mySegmentControl.frame = CGRect(x: 78, y: 517, width: 259, height: 30)
+        mySegmentControl.addTarget(self, action: #selector(selectedValue), for: .valueChanged)
+        view.addSubview(mySegmentControl)
+    }
```
code-style: отступы.

Без них принимать больше не буду.

Быстро отформатировать можно так:
- выделить все command+a
- отформатировать control+i

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -0,0 +1,79 @@
+            textFieldTwo.text = "Я изучаю UISegmentedControl"
+            self.view.addSubview(textFieldOne)
+            self.view.addSubview(textFieldTwo)
+            dismiss(animated: true, completion: nil)
```
dismiss не нужен.
dismiss закрывает контроллер, если тот был показан модально.
В данном случае контроллер показан с помощью UITabBarController, поэтому этот вызов избыточен.

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -0,0 +1,79 @@
+        view.addSubview(mySegmentControl)
+    }
+// создание сегмент контроллера
+    @objc func selectedValue(mySegmentControl: UISegmentedControl ) {
```
Данная функция может быть вызвана несколько раз, так как это обработчик события изменения сегмента.
Пользователь может сделать это очень много раз.

Получается так:
1) Пользователь нажал на сегмент.
2) Добавились вью из-за
```
            self.view.addSubview(textFieldOne)
             self.view.addSubview(textFieldTwo)
```
3) Пользователь нажал на сегмент.
4) Добавились вью из-за
```
            self.view.addSubview(textFieldOne)
             self.view.addSubview(textFieldTwo)
```
5) Пользователь нажал на сегмент.
6) Добавились вью из-за
```
            self.view.addSubview(textFieldOne)
             self.view.addSubview(textFieldTwo)
```
7) Пользователь нажал на сегмент.
8) Добавились вью из-за
```
            self.view.addSubview(textFieldOne)
             self.view.addSubview(textFieldTwo)
```

И так далее.

Создание и добавление вью необходимо производить один раз за все время жизни контроллера.
Функция viewDidLoad вызывается единожды за все время жизни контроллера.
Поэтому там будет ок создать и добавить все вьюхи.

По заданию требуется показывать/скрывать вьюхи.
Это можно сделать с помощью isHidden https://developer.apple.com/documentation/uikit/uiview/1622585-ishidden
В реализации `func selectedValue` необходимо только показывать/скрывать вью, соответствующие выбранному сегменту.

ios-review replied:
Про жизненный цикл контроллера по ссылке https://medium.com/good-morning-swift/ios-view-controller-life-cycle-2a0f02e74ff5

ios-review reviewed (changes requested):
В задании 1 есть критическая ошибка.
В задании 3 при изменении сегмента поведение не соответствует условию.

Также немного упростить алгоритмы + форматирование кода

TatyanaV24 committed:
Изменения ViewController

Внесла изменения в методы работы кнопок, теперь ошибки нет. Кнопки работают равноценно, переключая изображения.

TatyanaV24 committed:
Изменения в GalleryViewController

Создала модель, как Вы предложили. Однако, так и не смогла придумать, как именно выводить наименования в UILabel. Пыталась, как и с картинками, так и с UILabel вывести наименование, но всеопять начинает прыгать. Надеюсь поделитесь подсказкой, как лучше исправить этот недочет.

TatyanaV24 committed:
Правки SegmentViewController

Внесла правки по этой задаче. Сейчас она работает корректно. Спасибо за подсказку.
Пока не доконца разобралась с темой жизненный цикл контроллера, сегодня еще раз перечитаю статью и посмотрю пару уроков, чтобы окончательно усвоить.

TatyanaV24 committed:
Добавила доп.задачу в ветку

Задача решена не полностью.
Хочу получить Ваши комментарии, т.к. сейчас уперлась в стену и не понимаю, чего именно не замечаю.
С учетом условия задачи, рассуждаю так:
1. Пользователь наживает на кнопку "1", кнопка изменяет цвет и открывается первый контроллер.
2. Пользователь наживает на кнопку "2", кнопка изменяет цвет и открывается второй контроллер. При этом цвет кнопки "1" должен вернуться к первоначальному, а первый контроллер должен быть скрыт.
3. Пользователь наживает на кнопку "3", кнопка изменяет цвет и появляются два контрролера. При этом у кнопки "1" и "2" цвет сохраняется прежний.
Мои кнопки меняют цвет и выводят контроллеры, но мне так и не удалось научиться скрывать нужный контроллер и возвращать кнопкам цветам.
У меня такое ощущение, что я некорректно поняла условие задачи. Буду признательна за комментарии с Вашей стороны, чтобы двигаться дальше.
Спасибо

ios-review commented on HW8/HW8/BonusViewController.swift:
```diff
@@ -0,0 +1,68 @@
+
+import UIKit
+
+class BonusViewController: UIViewController {
```
Вынести в отдельный ПР.

Для каждого ДЗ и для каждого дополнительного создавать отдельную ветку + отдельный ПР.

ios-review replied:
И еще отдельный проект.

ios-review replied:
Обсудим в отдельном ПР.

TatyanaV24 replied:
Я правильно понимаю, сокращение "ПР" - обозначает новый проект?
Получается у меня будет две параллельные ветки. В одной будет решение задач по основному ДЗ, во второй ветке будут допущены. задачи.

ios-review replied:
ПР == пул реквест == pull request.
Для каждого задания, включая дополнительные, необходимо создавать отдельную ветку, отдельный проект, а после отдельный ПР.

> Получается у меня будет две параллельные ветки.

Веток может быть сколь угодно много.
Как пример, в моей работе в настоящий момент довольно много активных веток разработки, каждая из которых создана для каждой отдельной задачи.
Плюс каждый разработчик в команде также создает несколько отдельных веток для своих задач.

ios-review commented on HW8/HW8/GalleryViewController.swift:
```diff
@@ -0,0 +1,52 @@
+    var stepY = 0
+    var definitionOfPositonImage = 0
+    let catsImage = [UIImage(named:"1"), UIImage(named:"2"),UIImage(named:"3"),UIImage(named:"4")]
+    let catsName = ["Funny cat","Cute cat", "Smile rabbit", "Happy cat"]
```
Модель создана, но не используется.

Сейчас два массива, которые по сути связаны логически.
Изменив один массив, можно забыть изменить другой, что при выполнении приложения может привести к падению приложения, так как в цикле элементы массива достаются по индексу.
А получение элемента по индексу небезопасно, так как может не быть элемента с заданным индексом в массиве.
И это критическая ошибка, приводящая к падению приложения.

Использование массива моделей будет безопасно.
`let cats = [Cat(), Cat(), Cat()]`
И после использовать этот один массив с полными данными.
```
for cat in cats {
    // остальной код
    // cat.icon
    // cat.name
}
```

ios-review commented on HW8/HW8/ViewController.swift:
```diff
@@ -0,0 +1,35 @@
+        } else {
+            self.index = self.index-1
+            self.changeImageView.image = UIImage(named:catsImage[index])
+        }
```
```
self.index = self.index == 0 ? self.catsImage.count - 1 : self.index - 1
self.changeImageView.image = UIImage(named:catsImage[index])
```

ios-review commented on HW8/HW8/ViewController.swift:
```diff
@@ -0,0 +1,35 @@
+        } else {
+            self.index = self.index+1
+            self.changeImageView.image = UIImage(named:catsImage[index])
+        }
```
упростить

ios-review commented on HW8/HW8/ViewController.swift:
```diff
@@ -0,0 +1,35 @@
+
+import UIKit
+
+class ViewController: UIViewController {
```
<img width="444" alt="Screenshot 2020-11-26 at 17 00 03" src="https://user-images.githubusercontent.com/62285448/100337742-5cd1ed80-300a-11eb-9588-6c29db22b435.png">

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -0,0 +1,93 @@
+            //первый сегмент — зелёная view с двумя текстовыми полями;
+            self.view.backgroundColor = UIColor.green
+            self.view.addSubview(textFieldOne)
+            self.view.addSubview(textFieldTwo)
```
Эти вью будут добавляться несколько раз.
Фактически будут удаляться и тут же сразу добавляться.
Это неэффективно.

Достаточно их все сразу добавить в иерархию в viewDidLoad, а здесь только показывать/скрывать с помощью isHidden.

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -0,0 +1,93 @@
+            imageViewOne.isHidden = true
+            imageViewTwo.isHidden = true
+            yesButton.isHidden = true
+            noButton.isHidden = true
```
Если каждую вью поместить в контейнер, а здесь показывать/скрывать этот контейнер, то кода будет меньше.

```
View:
    View1: textField1,2
    View2: image1,2
    View3: button1,2

view1.isHidden = false
view2.isHidden = true
view3.isHidden = true
```

TatyanaV24 replied:
Вот этот момент, логически понимаю, но реально не представляю, как реализовать. Пыталась осуществить через массив и словари. Но они не работают с isHidden.
Также была попытка провести реализацию через enum, но также все не то.
Помогите, пожалуйста, порцией подсказок)

ios-review commented on HW8/HW8/GalleryViewController.swift:
```diff
@@ -0,0 +1,52 @@
+            view.addSubview(oneView)
+            oneView.addSubview(oneImage)
+            oneView.addSubview(oneLabel)
+        }
```
Алгоритм вычисления x и y можно еще упростить.
```
    override func viewDidLoad() {
        super.viewDidLoad()
        
        var x = 20
        var y = -250
        
        for (index, image) in images.enumerated() 
            if index % 2 == 0 {
                x = 20
                y += 270
            } else {
                x = 200
            }

            // остальной код
        }
    }
```

ios-review reviewed (changes requested):
Добавил подсказки)
Некоторые алгоритмы можно упростить.

По дополнительному заданию будем общаться в отдельном ПР.

TatyanaV24 committed:
Правки

Внесены основные правки по проекту.
Пересмотрела код во ViewController, упростила.
В GalleryViewController и SegmentViewController, также внесла правки по коду, и по циклам.
Единственный момент в GalleryViewController, никак немогу отладить корректную работу цикла. Не могу понять почему выводится только одна картинка. Подскажите, пожалуйста, где именно ошибка.
В SegmentViewController также не довела до конца, я оставила вопросе на github.

ios-review commented on HW8/HW8/GalleryViewController.swift:
```diff
@@ -0,0 +1,44 @@
+                } else {
+                    x = 200
+                }
+            }
```
А зачем нужен второй вложенный цикл?

Координаты x и y вычисляются на каждом шаге цикла для определенного объекта модели Cat.
```
var x = 20
var y = -250
for (index, cat) in cats.enumerated() {
    // Код вычисления x и y для текущей модели cat.
    // Код создания вьюх для текущей модели cat.
}
```

ios-review replied:
С двумя циклами получается так, что x и y для всех cat будет равен нижней правой вью, что наглядно видно при запуске программы.
Все вью создаются справа внизу.

Результат выполнения всегда можно посмотреть в View Debugger.
На скрине показал, как его открыть.
<img width="1400" alt="Screenshot 2020-11-29 at 17 41 08" src="https://user-images.githubusercontent.com/62285448/100539607-41651d80-326a-11eb-8cf8-1f37a5606901.png">

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        self.view.addSubview(noButton)
+        self.view.addSubview(imageViewOne)
+        self.view.addSubview(imageViewTwo)
+        
```
👍

Теперь, чтобы упростить показ/скрытие, будет удобно добавить каждую пару в отдельную вью.
```
let view1 = UIView()
view1.addSubview(textFieldOne)
view1.addSubview(textFieldTwo)

let view2 = UIView()
view2.addSubview(buttonOne)
view2.addSubview(buttonTwo)

view.addSubview(view1)
view.addSubview(view2)
```

Далее создать свойство для каждой из этих вью контейнеров: view1, view2, view3.
И показывать/скрывать именно их.

Для свойства isHidden действует правило: если вью скрыта, то и все ее подвью также скрыты.

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -0,0 +1,112 @@
+
+import UIKit
+
+class SegmentViewController: UIViewController {
```
<img width="1168" alt="Screenshot 2020-11-29 at 17 33 29" src="https://user-images.githubusercontent.com/62285448/100539696-cbad8180-326a-11eb-9c49-c8494fbd035b.png">

ios-review reviewed (changes requested):
Приложение будет выполнять все условия задачи, если поправить задание 2 (таба Item2).

Будет круто, если также упростите немного код.

TatyanaV24 committed:
окончательный вариант

Внесла необходимые правки по заданиям. Сейчас все работает.
Думаю, основная моя проблема в том, что я начинаю усложнять задачи и подсказки, которые получаю от Вас. Так стыдно, что я долго ходила кругами по заданию UISegmentController, ведь решение изначально было на поверхности. И я даже подумала сраз сделать именно так, но сочла это глупым вариантом и искала иную реализацию.

ios-review commented on HW8/HW8/GalleryViewController.swift:
```diff
@@ -16,26 +16,22 @@ class GalleryViewController: UIViewController {
+            } else {
+                x = 200
             }
+            let elArr = image
```
кажется, что в создании новой константы нет смысла.
```
oneImage.image = image.icon
oneLabel.image = image.name
```

TatyanaV24 replied:
Ой, это кусочек моих рассуждений, его не должно было быть.

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -41,70 +44,48 @@ class SegmentViewController: UIViewController {
-        imageViewTwo.isHidden = true
-        yesButton.isHidden = true
-        noButton.isHidden = true
+        view1.addSubview(textFieldOne)
+        view1.addSubview(textFieldTwo)
+        view2.addSubview(yesButton)
+        view2.addSubview(noButton)
+        view3.addSubview(imageViewOne)
+        view3.addSubview(imageViewTwo)
+        view.addSubview(view1)
+        view.addSubview(view2)
+        view.addSubview(view3)
+        view1.isHidden = true
+        view2.isHidden = true
+        view3.isHidden = true
```
👌

ios-review commented on HW8/HW8/SegmentViewController.swift:
```diff
@@ -41,70 +44,48 @@ class SegmentViewController: UIViewController {
             self.view.backgroundColor = UIColor.systemIndigo
-            imageViewOne.isHidden = false
-            imageViewTwo.isHidden = false
-            textFieldOne.isHidden = true
```
Более короткий алгоритм будет выглядеть так:
```
        for (index, view) in [view1, view2, view3].enumerated() {
            view.isHidden = index != mySegmentControl.selectedSegmentIndex
        }
```

А как можно сократить выбор цвета для view?

TatyanaV24 replied:
Добавила в код следующий вариант:
view.backgroundColor = mySegmentControl.selectedSegmentIndex == 0 ? UIColor.green : (mySegmentControl.selectedSegmentIndex == 1) ? UIColor.blue :  UIColor.systemIndigo
Это самый краткий вариант, который смогла придумать. Поделитесь своим?

ios-review reviewed (approved):
Задание готово👍
Молодец🔥

Можно мержить.
Упростить в качестве дополнительного задания.

TatyanaV24 committed:
Изменения по коду и сохранение

Внесла изменения в целях сокращения кода в контроллер SegmentViewController.
Заменила цикл по установке цвета background и цикл по скрытию/отображению элементов.
